File: lib/services/indicators.dart
import 'dart:math' as math;

import '../models/instrument.dart';

/// 量化指标纯计算工具集（无网络 IO，无 Tushare 依赖）。
/// 所有指标基于一段日线收盘序列；调用方负责拉取 [CandlePoint] 序列后传入。
class Indicators {
  Indicators._();

  // ── 基础工具 ────────────────────────────────────────────────────────────

  static List<double> _closes(List<CandlePoint> series) =>
      [for (final c in series) c.close];

  /// 收益率序列（pct_chg / 100），长度 = series.length - 1
  static List<double> dailyReturns(List<CandlePoint> series) {
    final out = <double>[];
    for (var i = 1; i < series.length; i++) {
      final prev = series[i - 1].close;
      if (prev == 0) {
        out.add(0);
      } else {
        out.add((series[i].close - prev) / prev);
      }
    }
    return out;
  }

  static double _mean(List<double> xs) {
    if (xs.isEmpty) return 0;
    return xs.reduce((a, b) => a + b) / xs.length;
  }

  static double _variance(List<double> xs) {
    if (xs.length < 2) return 0;
    final m = _mean(xs);
    var sum = 0.0;
    for (final x in xs) {
      final d = x - m;
      sum += d * d;
    }
    return sum / (xs.length - 1);
  }

  static double _stddev(List<double> xs) => math.sqrt(_variance(xs));

  static double _round(double v, [int digits = 4]) =>
      double.parse(v.toStringAsFixed(digits));

  // ── 收益 / 波动 / Sharpe / 最大回撤 ─────────────────────────────────────

  /// 区间累计收益率（最后/首日 - 1）
  static double cumulativeReturn(List<CandlePoint> series) {
    if (series.length < 2) return 0;
    final first = series.first.close;
    if (first == 0) return 0;
    return (series.last.close - first) / first;
  }

  /// 年化收益率（按 252 个交易日折算）
  static double annualizedReturn(List<CandlePoint> series) {
    if (series.length < 2) return 0;
    final r = cumulativeReturn(series);
    final years = (series.length - 1) / 252.0;
    if (years <= 0) return 0;
    return math.pow(1 + r, 1 / years) - 1;
  }

  /// 年化波动率（日收益率标准差 * sqrt(252)）
  static double annualizedVolatility(List<CandlePoint> series) {
    final rs = dailyReturns(series);
    if (rs.length < 2) return 0;
    return _stddev(rs) * math.sqrt(252);
  }

  /// 年化 Sharpe Ratio = (ann_ret - rf) / ann_vol
  static double sharpeRatio(List<CandlePoint> series, {double riskFree = 0.02}) {
    final ar = annualizedReturn(series);
    final av = annualizedVolatility(series);
    if (av == 0) return 0;
    return (ar - riskFree) / av;
  }

  /// 最大回撤：从历史峰值到谷底的最大跌幅，返回正数（如 0.235 表示 -23.5%）
  static MaxDrawdownResult maxDrawdown(List<CandlePoint> series) {
    if (series.length < 2) {
      return const MaxDrawdownResult(
          drawdown: 0, peakDate: null, troughDate: null);
    }
    var peak = series.first.close;
    DateTime peakDate = series.first.date;
    var maxDd = 0.0;
    DateTime? mddPeak;
    DateTime? mddTrough;
    for (final c in series) {
      if (c.close > peak) {
        peak = c.close;
        peakDate = c.date;
      }
      if (peak > 0) {
        final dd = (peak - c.close) / peak;
        if (dd > maxDd) {
          maxDd = dd;
          mddPeak = peakDate;
          mddTrough = c.date;
        }
      }
    }
    return MaxDrawdownResult(
      drawdown: _round(maxDd, 4),
      peakDate: mddPeak,
      troughDate: mddTrough,
    );
  }

  // ── 移动平均 / RSI / MACD ───────────────────────────────────────────────

  /// 简单移动平均 SMA
  static double? sma(List<CandlePoint> series, int n) {
    if (series.length < n || n <= 0) return null;
    final tail = _closes(series).sublist(series.length - n);
    return _round(_mean(tail), 4);
  }

  /// 指数加权移动平均 EMA
  static List<double> emaSeries(List<double> values, int n) {
    if (values.isEmpty || n <= 0) return const [];
    final k = 2 / (n + 1);
    final out = <double>[values.first];
    for (var i = 1; i < values.length; i++) {
      out.add(values[i] * k + out.last * (1 - k));
    }
    return out;
  }

  /// 相对强弱指数 RSI（默认 14 日）
  static double? rsi(List<CandlePoint> series, {int period = 14}) {
    if (series.length <= period) return null;
    var gains = 0.0;
    var losses = 0.0;
    for (var i = series.length - period; i < series.length; i++) {
      final diff = series[i].close - series[i - 1].close;
      if (diff > 0) {
        gains += diff;
      } else {
        losses -= diff;
      }
    }
    if (gains + losses == 0) return 50;
    final avgGain = gains / period;
    final avgLoss = losses / period;
    if (avgLoss == 0) return 100;
    final rs = avgGain / avgLoss;
    return _round(100 - 100 / (1 + rs), 2);
  }

  /// MACD：DIF = EMA(fast) - EMA(slow)；DEA = EMA(DIF, signal)；MACD = 2*(DIF-DEA)
  static MacdResult macd(
    List<CandlePoint> series, {
    int fast = 12,
    int slow = 26,
    int signal = 9,
  }) {
    if (series.length < slow + signal) {
      return const MacdResult(dif: null, dea: null, macd: null, cross: null);
    }
    final closes = _closes(series);
    final emaFast = emaSeries(closes, fast);
    final emaSlow = emaSeries(closes, slow);
    final dif = <double>[];
    for (var i = 0; i < closes.length; i++) {
      dif.add(emaFast[i] - emaSlow[i]);
    }
    final dea = emaSeries(dif, signal);
    final last = dif.length - 1;
    final macdVal = 2 * (dif[last] - dea[last]);
    String? cross;
    if (last >= 1) {
      final prevDif = dif[last - 1];
      final prevDea = dea[last - 1];
      if (prevDif <= prevDea && dif[last] > dea[last]) cross = 'golden';
      if (prevDif >= prevDea && dif[last] < dea[last]) cross = 'death';
    }
    return MacdResult(
      dif: _round(dif[last]),
      dea: _round(dea[last]),
      macd: _round(macdVal),
      cross: cross,
    );
  }

  // ── Beta / 相关性 ────────────────────────────────────────────────────────

  /// 收益率序列对齐：按日期裁剪到两边交集
  static (List<double>, List<double>) alignReturns(
      List<CandlePoint> a, List<CandlePoint> b) {
    final mapA = {for (final c in a) c.date: c.close};
    final mapB = {for (final c in b) c.date: c.close};
    final dates = mapA.keys.where(mapB.containsKey).toList()..sort();
    if (dates.length < 2) return (const [], const []);
    final ra = <double>[];
    final rb = <double>[];
    for (var i = 1; i < dates.length; i++) {
      final pa0 = mapA[dates[i - 1]]!;
      final pb0 = mapB[dates[i - 1]]!;
      if (pa0 == 0 || pb0 == 0) continue;
      ra.add((mapA[dates[i]]! - pa0) / pa0);
      rb.add((mapB[dates[i]]! - pb0) / pb0);
    }
    return (ra, rb);
  }

  /// Pearson 相关系数
  static double correlation(List<double> x, List<double> y) {
    if (x.length != y.length || x.length < 2) return 0;
    final mx = _mean(x);
    final my = _mean(y);
    var num = 0.0, dx2 = 0.0, dy2 = 0.0;
    for (var i = 0; i < x.length; i++) {
      final dx = x[i] - mx;
      final dy = y[i] - my;
      num += dx * dy;
      dx2 += dx * dx;
      dy2 += dy * dy;
    }
    final den = math.sqrt(dx2 * dy2);
    if (den == 0) return 0;
    return _round(num / den, 4);
  }

  /// Beta = cov(asset, benchmark) / var(benchmark)
  static BetaResult beta(List<CandlePoint> asset, List<CandlePoint> benchmark,
      {double riskFree = 0.02}) {
    final (ra, rb) = alignReturns(asset, benchmark);
    if (ra.length < 5) {
      return const BetaResult(beta: null, alpha: null, r2: null);
    }
    final ma = _mean(ra);
    final mb = _mean(rb);
    var cov = 0.0, varB = 0.0;
    for (var i = 0; i < ra.length; i++) {
      final dx = ra[i] - ma;
      final dy = rb[i] - mb;
      cov += dx * dy;
      varB += dy * dy;
    }
    cov /= ra.length - 1;
    varB /= ra.length - 1;
    if (varB == 0) {
      return const BetaResult(beta: null, alpha: null, r2: null);
    }
    final betaVal = cov / varB;
    final dailyRf = riskFree / 252;
    final alphaVal = (ma - dailyRf) - betaVal * (mb - dailyRf);
    final corr = correlation(ra, rb);
    return BetaResult(
      beta: _round(betaVal, 4),
      alpha: _round(alphaVal * 252, 4),
      r2: _round(corr * corr, 4),
    );
  }
}

class MaxDrawdownResult {
  const MaxDrawdownResult({
    required this.drawdown,
    required this.peakDate,
    required this.troughDate,
  });
  final double drawdown;
  final DateTime? peakDate;
  final DateTime? troughDate;
}

class MacdResult {
  const MacdResult({
    required this.dif,
    required this.dea,
    required this.macd,
    required this.cross,
  });
  final double? dif;
  final double? dea;
  final double? macd;
  final String? cross;
}

class BetaResult {
  const BetaResult({required this.beta, required this.alpha, required this.r2});
  final double? beta;
  final double? alpha;
  final double? r2;
}

File: backend/internal/ai/indicators/indicators.go
// Package indicators 是纯计算量化指标库（无网络 IO，无 Tushare 依赖），
// 输入是日线收盘序列，移植自 lib/services/indicators.dart 同语义函数。
package indicators

import (
	"math"
	"sort"
	"time"

	"github.com/sencloud/finme-backend/internal/ai/tushare"
)

// ── 基础工具 ────────────────────────────────────────────────────────────

// Closes 提取收盘序列副本。
func Closes(s []tushare.Candle) []float64 {
	out := make([]float64, len(s))
	for i, c := range s {
		out[i] = c.Close
	}
	return out
}

// DailyReturns 日收益率序列；长度 = len(s) - 1。
func DailyReturns(s []tushare.Candle) []float64 {
	out := make([]float64, 0, len(s))
	for i := 1; i < len(s); i++ {
		prev := s[i-1].Close
		if prev == 0 {
			out = append(out, 0)
			continue
		}
		out = append(out, (s[i].Close-prev)/prev)
	}
	return out
}

func mean(xs []float64) float64 {
	if len(xs) == 0 {
		return 0
	}
	sum := 0.0
	for _, x := range xs {
		sum += x
	}
	return sum / float64(len(xs))
}

func variance(xs []float64) float64 {
	if len(xs) < 2 {
		return 0
	}
	m := mean(xs)
	sum := 0.0
	for _, x := range xs {
		d := x - m
		sum += d * d
	}
	return sum / float64(len(xs)-1)
}

func stddev(xs []float64) float64 { return math.Sqrt(variance(xs)) }

// Round 四舍五入到 digits 位。
func Round(v float64, digits int) float64 {
	mul := math.Pow10(digits)
	return math.Round(v*mul) / mul
}

// ── 收益 / 波动 / Sharpe / 最大回撤 ─────────────────────────────────────

// CumulativeReturn 区间累计收益率。
func CumulativeReturn(s []tushare.Candle) float64 {
	if len(s) < 2 {
		return 0
	}
	first := s[0].Close
	if first == 0 {
		return 0
	}
	return (s[len(s)-1].Close - first) / first
}

// AnnualizedReturn 按 252 个交易日折算的年化。
func AnnualizedReturn(s []tushare.Candle) float64 {
	if len(s) < 2 {
		return 0
	}
	r := CumulativeReturn(s)
	years := float64(len(s)-1) / 252.0
	if years <= 0 {
		return 0
	}
	return math.Pow(1+r, 1/years) - 1
}

// AnnualizedVolatility 日收益率标准差 * sqrt(252)。
func AnnualizedVolatility(s []tushare.Candle) float64 {
	rs := DailyReturns(s)
	if len(rs) < 2 {
		return 0
	}
	return stddev(rs) * math.Sqrt(252)
}

// SharpeRatio 年化 Sharpe Ratio。
func SharpeRatio(s []tushare.Candle, riskFree float64) float64 {
	ar := AnnualizedReturn(s)
	av := AnnualizedVolatility(s)
	if av == 0 {
		return 0
	}
	return (ar - riskFree) / av
}

// MaxDrawdownResult 包含最大回撤和峰/谷日期。
type MaxDrawdownResult struct {
	Drawdown   float64
	PeakDate   time.Time
	TroughDate time.Time
	Has        bool
}

// MaxDrawdown 返回区间最大回撤（正数，例 0.235 = -23.5%）。
func MaxDrawdown(s []tushare.Candle) MaxDrawdownResult {
	if len(s) < 2 {
		return MaxDrawdownResult{}
	}
	peak := s[0].Close
	peakDate := parseTradeDate(s[0].TradeDate)
	maxDd := 0.0
	var mddPeak, mddTrough time.Time
	has := false
	for _, c := range s {
		if c.Close > peak {
			peak = c.Close
			peakDate = parseTradeDate(c.TradeDate)
		}
		if peak > 0 {
			dd := (peak - c.Close) / peak
			if dd > maxDd {
				maxDd = dd
				mddPeak = peakDate
				mddTrough = parseTradeDate(c.TradeDate)
				has = true
			}
		}
	}
	return MaxDrawdownResult{
		Drawdown:   Round(maxDd, 4),
		PeakDate:   mddPeak,
		TroughDate: mddTrough,
		Has:        has,
	}
}

// ── 移动平均 / RSI / MACD ───────────────────────────────────────────────

// SMA 简单移动平均；长度不足返回 (0, false)。
func SMA(s []tushare.Candle, n int) (float64, bool) {
	if n <= 0 || len(s) < n {
		return 0, false
	}
	closes := Closes(s)
	tail := closes[len(closes)-n:]
	return Round(mean(tail), 4), true
}

// EMASeries 指数加权移动平均整段序列。
func EMASeries(values []float64, n int) []float64 {
	if len(values) == 0 || n <= 0 {
		return nil
	}
	k := 2.0 / (float64(n) + 1.0)
	out := make([]float64, len(values))
	out[0] = values[0]
	for i := 1; i < len(values); i++ {
		out[i] = values[i]*k + out[i-1]*(1-k)
	}
	return out
}

// RSI 相对强弱指数（默认 14 日）。
func RSI(s []tushare.Candle, period int) (float64, bool) {
	if period <= 0 || len(s) <= period {
		return 0, false
	}
	gains, losses := 0.0, 0.0
	for i := len(s) - period; i < len(s); i++ {
		diff := s[i].Close - s[i-1].Close
		if diff > 0 {
			gains += diff
		} else {
			losses -= diff
		}
	}
	if gains+losses == 0 {
		return 50, true
	}
	avgGain := gains / float64(period)
	avgLoss := losses / float64(period)
	if avgLoss == 0 {
		return 100, true
	}
	rs := avgGain / avgLoss
	return Round(100-100/(1+rs), 2), true
}

// MacdResult 是 MACD 三值与金/死叉信号。
type MacdResult struct {
	Dif   *float64
	Dea   *float64
	Macd  *float64
	Cross string
}

// MACD 计算最后一日的 DIF/DEA/MACD 与是否金叉/死叉。
func MACD(s []tushare.Candle, fast, slow, signal int) MacdResult {
	if len(s) < slow+signal {
		return MacdResult{}
	}
	closes := Closes(s)
	emaFast := EMASeries(closes, fast)
	emaSlow := EMASeries(closes, slow)
	dif := make([]float64, len(closes))
	for i := range closes {
		dif[i] = emaFast[i] - emaSlow[i]
	}
	dea := EMASeries(dif, signal)
	last := len(dif) - 1
	macdVal := 2 * (dif[last] - dea[last])
	cross := ""
	if last >= 1 {
		prevDif, prevDea := dif[last-1], dea[last-1]
		if prevDif <= prevDea && dif[last] > dea[last] {
			cross = "golden"
		}
		if prevDif >= prevDea && dif[last] < dea[last] {
			cross = "death"
		}
	}
	d := Round(dif[last], 4)
	a := Round(dea[last], 4)
	m := Round(macdVal, 4)
	return MacdResult{Dif: &d, Dea: &a, Macd: &m, Cross: cross}
}

// ── Beta / 相关性 ────────────────────────────────────────────────────────

// AlignReturns 把两段日线按日期交集对齐成日收益率序列。
func AlignReturns(a, b []tushare.Candle) (ra, rb []float64) {
	mapA := make(map[string]float64, len(a))
	mapB := make(map[string]float64, len(b))
	for _, c := range a {
		mapA[c.TradeDate] = c.Close
	}
	for _, c := range b {
		mapB[c.TradeDate] = c.Close
	}
	dates := make([]string, 0, len(mapA))
	for d := range mapA {
		if _, ok := mapB[d]; ok {
			dates = append(dates, d)
		}
	}
	sort.Strings(dates)
	if len(dates) < 2 {
		return nil, nil
	}
	for i := 1; i < len(dates); i++ {
		pa0 := mapA[dates[i-1]]
		pb0 := mapB[dates[i-1]]
		if pa0 == 0 || pb0 == 0 {
			continue
		}
		ra = append(ra, (mapA[dates[i]]-pa0)/pa0)
		rb = append(rb, (mapB[dates[i]]-pb0)/pb0)
	}
	return ra, rb
}

// Correlation Pearson 相关系数。
func Correlation(x, y []float64) float64 {
	if len(x) != len(y) || len(x) < 2 {
		return 0
	}
	mx := mean(x)
	my := mean(y)
	num, dx2, dy2 := 0.0, 0.0, 0.0
	for i := range x {
		dx := x[i] - mx
		dy := y[i] - my
		num += dx * dy
		dx2 += dx * dx
		dy2 += dy * dy
	}
	den := math.Sqrt(dx2 * dy2)
	if den == 0 {
		return 0
	}
	return Round(num/den, 4)
}

// BetaResult 包含 beta、年化 alpha、r²。
type BetaResult struct {
	Beta  *float64
	Alpha *float64
	R2    *float64
}

// Beta 估算资产相对基准的 Beta、年化 Alpha 和 R²。
func Beta(asset, benchmark []tushare.Candle, riskFree float64) BetaResult {
	ra, rb := AlignReturns(asset, benchmark)
	if len(ra) < 5 {
		return BetaResult{}
	}
	ma := mean(ra)
	mb := mean(rb)
	cov, varB := 0.0, 0.0
	for i := range ra {
		dx := ra[i] - ma
		dy := rb[i] - mb
		cov += dx * dy
		varB += dy * dy
	}
	cov /= float64(len(ra) - 1)
	varB /= float64(len(ra) - 1)
	if varB == 0 {
		return BetaResult{}
	}
	betaVal := cov / varB
	dailyRf := riskFree / 252.0
	alphaVal := (ma - dailyRf) - betaVal*(mb-dailyRf)
	corr := Correlation(ra, rb)
	b := Round(betaVal, 4)
	a := Round(alphaVal*252, 4)
	r := Round(corr*corr, 4)
	return BetaResult{Beta: &b, Alpha: &a, R2: &r}
}

func parseTradeDate(yyyymmdd string) time.Time {
	if len(yyyymmdd) != 8 {
		return time.Time{}
	}
	t, err := time.Parse("20060102", yyyymmdd)
	if err != nil {
		return time.Time{}
	}
	return t
}
